Documentation organization and additions

chris-martin opened the pull request:
I ended up doing more than was discussed in #21, hope this is okay...

The module comment is broken up into smaller blocks to incorporate it into the table of contents, and a few bits of it are moved down to relevant sections within the API documentation.

Modified the "test your understanding" example a bit:

- Avoid use of `select` because it isn't in the list of core functions
- Concat strings with `<|>` to show it works with regular lists too
- Print a blank line between prompts so it's easier to follow when it runs

`mapM` discussion is shortened

New documentation:

- doctest on `fold`
- `empty`/`pure`/`return` reiterated in "Constructing" section
- Add `filter` definition from #3 given as an example
- Add "Concatenation" section mentioning `<|>`, `asum`, and `join`
- Add "Pairwise combination" section explaining `<>` and its relationship with `>>=`
- "Repetition" section gives `cycle` and `replicate` example functions, including the ones from #3 and more

chris-martin committed:
Documentation organization and additions

The module comment is broken up into smaller blocks to incorporate it
into the table of contents, and a few bits of it are moved down to
relevant sections within the API documentation.

Modified the "test your understanding" example a bit:
  - Concat strings with <|> to show it works with regular lists too
  - Print a blank line between prompts so it's easier to follow when it
    runs

Added `select` to the list of core functions so that it will be
introduced before it is used in the exercise

`mapM` discussion is shortened

New documentation:
  - doctest on `fold`
  - empty/pure/return reiterated in "Constructing" section
  - Add `filter` definition from #3 given as an example
  - Add "Concatenation" section mentioning `<|>`, `asum`, and `join`
  - Add "Pairwise combination" section explaining `<>` and its
    relationship with `>>=`
  - "Repetition" section gives `cycle` and `replicate` example functions,
    including the ones from #3 and more

Gabriella439 commented on src/List/Transformer.hs:
```diff
@@ -104,129 +184,69 @@
-import Prelude hiding (drop, dropWhile, pred, take, takeWhile, zip)
+> strings :: ListT IO String
+> strings = do
+>     asum (repeat (return ()))
```
```suggestion
>     select (repeat ())
```

Gabriella439 commented on src/List/Transformer.hs:
```diff
@@ -104,129 +184,69 @@
-    , MonadIO(..)
-    , Alternative(..)
-    ) where
+@
```
I think this example should combine all three code blocks since they form one file

Gabriella439 commented on src/List/Transformer.hs:
```diff
@@ -7,19 +7,111 @@
+    , runListT
+    , fold
+    , foldM
+      -- $pleaseStream
```
I think I would move the `$pleaseStream` notice to the beginning of the "Consuming" section (and change "above examples" to "following example"), otherwise it's easy to miss

Gabriella439 reviewed (approved):
Looks great!  Just a few comments

chris-martin commented:
- [x] Moved the `$pleaseStream` to the top of the section
- [x] Consolidated the exercise into one code block
- [x] Changed the exercise line back to use `select` instead of `asum`, like you had it before. I added `select` to the list of "most important operations" so that it can be introduced before it is used in the exercise; I think it belongs there anyway.

(By the way, I've also raised a [proposal](https://github.com/haskell/core-libraries-committee/issues/56) to add `select` to `Control.Applicative`, would welcome any comments.)

Gabriella439 commented:
Sorry, I got distracted by the proposal discussion and forgot about this PR 😅
